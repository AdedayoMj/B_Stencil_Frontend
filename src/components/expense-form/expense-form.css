:host {
  display: block;
}
form {
  max-width: 400px;
  margin: 0 auto;
}

.form-group {
  margin-bottom: 1rem;
}

label {
  display: block;
  font-weight: bold;
  margin-bottom: 0.5rem;
  margin-left: 0.5rem;
}

input,
button {
  display: block;
  width: 100%;
  padding: 0.5rem;
  font-size: 1rem;
  border-radius: 4px;
  border: 1px solid #ccc;
  margin: 8px;
}

button {
  background-color: var(--app-button-color2);
  color: #fff;
  cursor: pointer;
  padding: 0.5rem 2rem 0.5rem 2rem;
  transition: background-color 0.3s;
  width: auto;
  margin: 0 auto;
}

button:hover {
  background-color: #0b7a94;
}

.error input {
  border-color: red;
}

.error-message {
  color: red;
  font-size: 0.8rem;
  margin-top: 0.5rem;
  margin-left: 0.5rem;
}
