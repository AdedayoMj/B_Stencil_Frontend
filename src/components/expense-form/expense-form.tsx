import { Component, State, h, Element, Prop } from '@stencil/core';
import { CreateExpenseData } from '../../types';
import { expenseStore } from '../../store/expense-store';

@Component({
  tag: 'expense-form',
  styleUrl: 'expense-form.css',
  shadow: true,
})
export class ExpenseForm {
  @Prop() showForm = false;
  @State() action: string = expenseStore.getState().action;
  @State() _id: string = expenseStore.getState().currentId;

  @State() formData: CreateExpenseData = {
    date: '',
    description: '',
    amount: null,
  };

  @State() formErrors: { [key: string]: string } = {};
  @Element() element: HTMLElement;

  connectedCallback() {
    if (this.action === 'update' && this._id) {
      const expense = expenseStore.getExpenseById(this._id);
      if (expense) {
        this.formData = {
          date: expense.date,
          description: expense.description,
          amount: expense.amount,
        };
      } else {
        throw new Error(`Expense not found for _id: ${this._id}`);
      }
    }
    this.element.addEventListener('modalClose', () => {
      this.resetForm();
    });
  }

  handleInput(event: Event) {
    const inputElement = event.target as HTMLInputElement;
    const { name, value } = inputElement;

    this.formData = {
      ...this.formData,
      [name]: value,
    };

    this.formErrors[name] = '';
  }

  async handleSubmit(event: Event) {
    event.preventDefault();

    // Validate form fields
    const { date, description, amount } = this.formData;
    let formIsValid = true;

    if (!date) {
      this.formErrors.date = 'Date is required';
      formIsValid = false;
    }

    if (!description) {
      this.formErrors.description = 'Expense Details is required';
      formIsValid = false;
    }

    if (!amount) {
      this.formErrors.amount = 'Amount is required';
      formIsValid = false;
    }

    if (formIsValid) {
      if (this.action === 'create') {
        await expenseStore.createExpense(this.formData);
      }

      if (this.action === 'update') {
        await expenseStore.updateExpenseApi(this._id, this.formData);
      }

      await expenseStore.fetchExpenses();
      this.resetForm();
    }
  }

  resetForm() {
    this.formData = {
      date: '',
      description: '',
      amount: null,
    };
    expenseStore.setCurrentId('');
    expenseStore.setShowForm(false);
    expenseStore.setAction('create');
  }

  render() {
    return (
      <app-modal isOpen={this.showForm}>
        <form class="form" onSubmit={e => this.handleSubmit(e)}>
          <div class="form-group">
            <label htmlFor="date">Date</label>
            <input type="date" id="date" value={this.formData.date} name="date" onInput={e => this.handleInput(e)} class={this.formErrors.date ? 'error' : ''} />
            {this.formErrors.date && <div class="error-message">{this.formErrors.date}</div>}
          </div>
          <div class="form-group">
            <label htmlFor="description">Expense Details</label>
            <textarea
              id="description"
              value={this.formData.description}
              name="description"
              rows={5}
              onInput={e => this.handleInput(e)}
              class={this.formErrors.description ? 'error' : ''}
            />
            {this.formErrors.description && <div class="error-message">{this.formErrors.description}</div>}
          </div>
          <div class="form-group">
            <label htmlFor="amount">Amount</label>
            <input type="number" id="amount" value={this.formData.amount} name="amount" onInput={e => this.handleInput(e)} class={this.formErrors.amount ? 'error' : ''} />
            {this.formErrors.amount && <div class="error-message">{this.formErrors.amount}</div>}
          </div>
          <div class="form-group">
            <button type="submit">Submit</button>
          </div>
        </form>
      </app-modal>
    );
  }
}
