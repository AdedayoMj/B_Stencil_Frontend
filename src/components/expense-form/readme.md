# expense-form

<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description | Type      | Default |
| ---------- | ----------- | ----------- | --------- | ------- |
| `showForm` | `show-form` |             | `boolean` | `false` |


## Dependencies

### Used by

 - [app-root](../app-root)

### Depends on

- [app-modal](../app-modal)

### Graph
```mermaid
graph TD;
  expense-form --> app-modal
  app-root --> expense-form
  style expense-form fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
