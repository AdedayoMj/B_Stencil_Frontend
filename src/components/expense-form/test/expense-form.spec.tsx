import { newSpecPage } from '@stencil/core/testing';
import { ExpenseForm } from '../expense-form';
// import { expenseStore } from '../../store/expense-store';

describe('ExpenseForm', () => {
  it('should render expense form correctly', async () => {
    const page = await newSpecPage({
      components: [ExpenseForm],
      html: `<expense-form></expense-form>`,
    });

    expect(page.root).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should update form data on input change', async () => {
    const page = await newSpecPage({
      components: [ExpenseForm],
      html: `<expense-form show-form="true"></expense-form>`,
    });

    await page.waitForChanges();

    const expenseForm = page.rootInstance as ExpenseForm;

    const inputElement = page.root.shadowRoot.querySelector('input[name="date"]') as HTMLInputElement;
    inputElement.value = '2023-07-09';
    inputElement.dispatchEvent(new Event('input'));

    expect(expenseForm.formData.date).toBe('2023-07-09');
  });

  it('should display form errors correctly', async () => {
    const page = await newSpecPage({
      components: [ExpenseForm],
      html: `<expense-form show-form="true"></expense-form>`,
    });
    await page.waitForChanges();
    const expenseForm = page.rootInstance as ExpenseForm;

    const formElement = page.root.shadowRoot.querySelector('.form');
    formElement.dispatchEvent(new Event('submit'));

    await page.waitForChanges();

    expect(expenseForm.formErrors.date).toBe('Date is required');
    expect(expenseForm.formErrors.description).toBe('Expense Details is required');
    expect(expenseForm.formErrors.amount).toBe('Amount is required');
  });
});
