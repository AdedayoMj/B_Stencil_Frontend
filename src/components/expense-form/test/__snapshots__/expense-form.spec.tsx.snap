// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseForm should render expense form correctly 1`] = `
<expense-form>
  <mock:shadow-root>
    <app-modal>
      <form class="form">
        <div class="form-group">
          <label htmlfor="date">
            Date
          </label>
          <input id="date" name="date" type="date" value="">
        </div>
        <div class="form-group">
          <label htmlfor="description">
            Expense Details
          </label><textarea id="description" name="description" rows="5" value=""></textarea>
        </div>
        <div class="form-group">
          <label htmlfor="amount">
            Amount
          </label>
          <input id="amount" name="amount" type="number">
        </div>
        <div class="form-group">
          <button type="submit">
            Submit
          </button>
        </div>
      </form>
    </app-modal>
  </mock:shadow-root>
</expense-form>
`;
